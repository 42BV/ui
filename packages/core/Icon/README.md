# @42.nl/ui-core-icon

## Installation

Install the package in your project directory with:

```sh
// with npm
npm install @42.nl/ui-core-icon

// with yarn
yarn add @42.nl/ui-core-icon
```
