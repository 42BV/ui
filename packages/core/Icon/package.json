{
  "name": "@42.nl/ui-core-icon",
  "version": "0.0.1",
  "description": "",
  "main": "lib/index.js",
  "private": false,
  "scripts": {
    "test": "jest --config=jest.config.js",
    "build": "tsc",
    "lint": "eslint src --ext js,jsx,ts,tsx --max-warnings=0",
    "clean": "rm -rf lib"
  },
  "files": [
    "lib/*"
  ],
  "dependencies": {
    "@42.nl/ui-types": "0.0.1",
    "classnames": "2.2.6"
  },
  "peerDependencies": {
    "react": "^16.9.0",
    "react-dom": "^16.9.0",
    "reactstrap": "^8.0.1"
  },
  "devDependencies": {
    "reactstrap": "8.0.1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "publishConfig": {
    "access": "public"
  }
}
