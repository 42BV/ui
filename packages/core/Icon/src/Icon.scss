// @remove-on-build-start
@import '~@42.nl/ui-core-styling/lib/bootstrap';
@import '~@42.nl/ui-core-styling/lib/icons';
// @remove-on-build-end

/**
* Make material-icons within Bootstrap great again!
*/
.input-group {
  &-append,
  &-prepend {
    i {
      @extend .input-group-text;
    }

    .btn > i.material-icons {
      border: 0;
      background: none;
      color: #fff;
      padding: 0;
    }
  }
}

.nav-link {
  i.material-icons {
    vertical-align: middle;
  }
}
