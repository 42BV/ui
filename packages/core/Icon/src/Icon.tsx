import React from 'react';
import classNames from 'classnames';

import IconType from './types';

interface Props {
  /**
   * The material icon you want to render.
   *
   * @type {IconType}
   * @memberof Props
   */
  icon: IconType;

  /**
   * Optional color you want the Icon to have. Support bootstrap colors.
   *
   * @type {string}
   * @memberof Props
   */
  color?: string;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   *
   * @type {string}
   * @memberof Props
   */
  className?: string;

  /**
   * Optional id for when you need it for tooltips we sometimes do need it.
   *
   * @type {string}
   * @memberof Props
   */
  id?: string;

  /**
   * Optional onClick event for when the icon is clicked.
   */
  onClick?: (event: React.MouseEvent<HTMLElement>) => void;
}

/**
 * The Icon is a small wrapper around a material design icon.
 *
 * Useful for making sure your icon is typesafe via TypeScript.
 *
 * @export
 * @param {Props} props
 * @returns
 */
export default function Icon({ className, onClick, icon, id, color }: Props) {
  const colorCssClass = color !== undefined ? `text-${color}` : undefined;
  const classes = classNames(className, 'material-icons', colorCssClass, {
    clickable: onClick
  });

  return (
    <i id={id} onClick={onClick} className={classes}>
      {icon}
    </i>
  );
}
