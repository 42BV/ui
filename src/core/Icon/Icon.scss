/**
* Make material-icons within Bootstrap great again!
*/
.input-group {
  &-append,
  &-prepend {
    i {
      @extend .input-group-text;
    }

    .btn > i.material-icons {
      border: 0;
      background: none;
      color: #fff;
      padding: 0;
    }
  }
}

i.material-icons {
  user-select: none;
}

.nav-link {
  i.material-icons {
    vertical-align: middle;
  }
}
