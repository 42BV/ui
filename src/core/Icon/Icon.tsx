import React from 'react';
import classNames from 'classnames';

import { Color, useHover } from '../..';
import IconType from './types';

export type Props = {
  /**
   * The material icon you want to render.
   */
  icon: IconType;

  /**
   * Optional color you want the Icon to have.
   */
  color?: Color;

  /**
   * Optional color you want the Icon to have when it is hovered.
   * When `disabled` is true, hoverColor does not work.
   * This will only be used when onClick is defined.
   */
  hoverColor?: Color;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;

  /**
   * Optional id when using controlled tooltips.
   */
  id?: string;

  /**
   * Optional onClick event for when the Icon is clicked.
   */
  onClick?: (event: React.MouseEvent<HTMLElement>) => void;

  /**
   * Optionally whether the button is disabled
   *
   * Defaults to `false`
   */
  disabled?: boolean;

  /**
   * Optionally the size of the icon in pixels.
   *
   * Defaults to `24px`
   */
  size?: number;
};

/**
 * The Icon is a small wrapper around a material design icon.
 *
 * Useful for making sure your icon is typesafe via TypeScript.
 */
export default function Icon({
  className,
  onClick,
  icon,
  id,
  color,
  hoverColor,
  disabled,
  size
}: Props) {
  const [hover, hoverEvents] = useHover();

  const colorCssClasses = {};
  if (color) {
    colorCssClasses[`text-${color}`] = (hoverColor && !hover) || disabled || onClick === undefined;
  }
  if (hoverColor) {
    colorCssClasses[`text-${hoverColor}`] = hover && !disabled && onClick !== undefined;
  }

  const classes = classNames(
    'icon',
    className,
    'material-icons',
    colorCssClasses,
    {
      clickable: !disabled && onClick,
      'icon--disabled': disabled
    }
  );

  const style = size ? { fontSize: size } : undefined;

  return (
    <i
      id={id}
      style={style}
      onClick={(event) => {
        if (!disabled && onClick) {
          onClick(event);
        }
      }}
      className={classes}
      {...hoverEvents}
    >
      {icon}
    </i>
  );
}
