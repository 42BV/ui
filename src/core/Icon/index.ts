export { Icon } from './Icon';
export { IconType, icons } from './types';
