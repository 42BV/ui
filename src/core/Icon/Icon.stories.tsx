import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { Icon } from './Icon';
import { Button } from '../Button/Button';

storiesOf('core/Icons', module)
  .addParameters({ component: Icon })
  .add('available icons', () => {
    return (
      <>
        <p>You can find the available icons on Google Fonts</p>
        <Button
          onClick={() =>
            window.open('https://fonts.google.com/icons', '_blank')
          }
        >
          Go to Icons on Google Fonts
        </Button>
      </>
    );
  })
  .add('examples', () => {
    return (
      <div className="d-flex flex-column align-items-center">
        <Icon icon="child_care" color="danger" />
        <Icon icon="adb" color="info" />
        <Icon icon="restaurant" color="muted" />
        <Icon icon="train" color="primary" />
        <Icon icon="wb_sunny" color="warning" />
        <Icon icon="info" size={144} />
        <Icon icon="info" size={20} />
        <Icon icon="drafts" />
        <Icon icon="drafts" disabled={true} />
        <Icon icon="timer_3" onClick={action('timer clicked')} />
        <Icon icon="home" disabled={true} onClick={action('home clicked')} />
      </div>
    );
  })
  .add('hover color', () => {
    return (
      <div className="d-flex flex-column align-items-center">
        <Icon
          icon="child_care"
          color="secondary"
          hoverColor="danger"
          onClick={action('child care clicked')}
        />
        <Icon
          icon="train"
          color="primary"
          hoverColor="success"
          onClick={action('train clicked')}
        />
        <Icon
          icon="home"
          disabled={true}
          color="primary"
          hoverColor="success"
          onClick={action('home clicked')}
        />
      </div>
    );
  });
