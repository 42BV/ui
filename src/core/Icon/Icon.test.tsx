import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import { Icon } from './Icon';
import * as components from './index';

describe('index', () => {
  it('should contain exports', () => {
    expect(components).toMatchSnapshot();
  });
});

describe('Icon', () => {
  describe('ui', () => {
    test('default', () => {
      const { container } = render(<Icon icon="alarm" />);
      expect(container).toMatchSnapshot();
    });

    test('with id', () => {
      const { container } = render(<Icon icon="alarm" id="1337" />);
      expect(container).toMatchSnapshot();
    });

    test('with color', () => {
      const { container } = render(
        <Icon icon="alarm" id="1337" color="danger" />
      );
      expect(container.firstChild).toHaveClass('text-danger');
    });

    test('with className', () => {
      const { container } = render(
        <Icon icon="alarm" className="text-danger" />
      );
      expect(container.firstChild).toHaveClass('text-danger');
    });

    test('is clickable', () => {
      const { container } = render(
        <Icon icon="alarm" onClick={() => undefined} />
      );
      expect(container.firstChild).toHaveClass('clickable');
    });

    test('is not clickable', () => {
      const { container } = render(<Icon icon="alarm" onClick={undefined} />);
      expect(container.firstChild).not.toHaveClass('clickable');
    });

    test('is disabled', () => {
      const { container } = render(
        <Icon icon="alarm" onClick={() => undefined} disabled={true} />
      );
      expect(container.firstChild).toHaveClass('icon--disabled');
    });

    test('is enabled', () => {
      const { container } = render(<Icon icon="alarm" disabled={false} />);
      expect(container.firstChild).not.toHaveClass('icon--disabled');
    });

    test('with size', () => {
      const { container } = render(<Icon icon="alarm" size={144} />);
      expect(container.firstChild).toHaveStyle({ fontSize: 144 });
    });

    test('with variant', () => {
      const { container } = render(<Icon icon="alarm" variant="outlined" />);
      expect(container.firstChild).toHaveClass('material-icons__outlined');
    });
  });

  describe('hoverColor', () => {
    it('should use hoverColor when onClick and hoverColor are defined', async () => {
      expect.assertions(1);
      const { container } = render(
        <Icon
          icon="alarm"
          color="secondary"
          hoverColor="primary"
          onClick={jest.fn()}
        />
      );
      await userEvent.hover(screen.getByText('alarm'));
      expect(container.firstChild).toHaveClass('text-primary');
    });

    it('should not use hoverColor when onClick is not defined', async () => {
      expect.assertions(1);
      const { container } = render(
        <Icon icon="alarm" color="secondary" hoverColor="primary" />
      );
      await userEvent.hover(screen.getByText('alarm'));
      expect(container.firstChild).toHaveClass('text-secondary');
    });

    it('should not use hoverColor when onClick and hoverColor are defined but icon is disabled', async () => {
      expect.assertions(1);
      const { container } = render(
        <Icon
          icon="alarm"
          color="secondary"
          hoverColor="primary"
          onClick={jest.fn()}
          disabled={true}
        />
      );
      await userEvent.hover(screen.getByText('alarm'));
      expect(container.firstChild).toHaveClass('text-secondary');
    });
  });

  describe('events', () => {
    it('should call the "onClick" event when the icon is clicked', () => {
      const onClickSpy = jest.fn();

      render(<Icon icon="alarm" onClick={onClickSpy} />);

      fireEvent.click(screen.getByText('alarm'));

      expect(onClickSpy).toHaveBeenCalledTimes(1);
    });
    it('should call the "onClick" event when the icon has focus and the enter key is pressed', () => {
      const onClickSpy = jest.fn();

      render(<Icon icon="alarm" onClick={onClickSpy} />);

      fireEvent.keyUp(screen.getByText('alarm'), { key: 'Enter' });

      expect(onClickSpy).toHaveBeenCalledTimes(1);
    });

    it('should call the "onClick" event when the icon is clicked and is explicitly enabled', () => {
      const onClickSpy = jest.fn();

      render(<Icon icon="alarm" onClick={onClickSpy} disabled={false} />);

      fireEvent.click(screen.getByText('alarm'));

      expect(onClickSpy).toHaveBeenCalledTimes(1);
    });

    it('should not call the "onClick" event when the icon is clicked but is disabled', () => {
      const onClickSpy = jest.fn();

      render(<Icon icon="alarm" onClick={onClickSpy} disabled={true} />);

      fireEvent.click(screen.getByText('alarm'));

      expect(onClickSpy).toHaveBeenCalledTimes(0);
    });
  });
});
