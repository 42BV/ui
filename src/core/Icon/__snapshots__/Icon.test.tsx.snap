// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon ui default: Icon => ui => default 1`] = `
<i
  className="icon material-icons"
  onClick={[Function]}
>
  alarm
</i>
`;

exports[`Icon ui is clickable: Icon => ui => is clickable 1`] = `
<i
  className="icon material-icons clickable"
  onClick={[Function]}
>
  alarm
</i>
`;

exports[`Icon ui is disabled: Icon => ui => is disabled 1`] = `
<i
  className="icon material-icons icon--disabled"
  onClick={[Function]}
>
  alarm
</i>
`;

exports[`Icon ui is enabled: Icon => ui => is enabled 1`] = `
<i
  className="icon material-icons"
  onClick={[Function]}
>
  alarm
</i>
`;

exports[`Icon ui is not clickable: Icon => ui => is not clickable 1`] = `
<i
  className="icon material-icons"
  onClick={[Function]}
>
  alarm
</i>
`;

exports[`Icon ui with className: Icon => ui => with className 1`] = `
<i
  className="icon text-danger material-icons"
  onClick={[Function]}
>
  alarm
</i>
`;

exports[`Icon ui with color: Icon => ui => with color 1`] = `
<i
  className="icon material-icons text-danger"
  id="1337"
  onClick={[Function]}
>
  alarm
</i>
`;

exports[`Icon ui with id: Icon => ui => with id 1`] = `
<i
  className="icon material-icons"
  id="1337"
  onClick={[Function]}
>
  alarm
</i>
`;

exports[`index should contain exports 1`] = `
Object {
  "Icon": [Function],
}
`;
